pkcs7/src/builder.rs

```diff
@@ -0,0 +1,541 @@
+            }
+        } else {
+            // Add signed attributes with signing time attribute and content type attribute
+            let mut signed_attributes = SignedAttributes::new();
+            signed_attributes.add(create_signing_time_attribute()?)?;
+            signer_info.signed_attributes = Some(signed_attributes);
+        }
+
+        // Add digest attribute to (to be) signed attributes
+        let signed_attributes = signer_info
+            .signed_attributes
+            .as_mut()
+            .expect("Signed attributes must be present.");
+        signed_attributes.add(create_message_digest_attribute(&message_digest)?)?;
+
+        // The content-type attribute type specifies the content type of the
+        // ContentInfo within signed-data or authenticated-data.  The content-
+        // type attribute type MUST be present whenever signed attributes are
+        // present in signed-data or authenticated attributes present in
+        // authenticated-data.  The content-type attribute value MUST match the
+        // encapContentInfo eContentType value in the signed-data or
+        // authenticated-data.
+        let e_content_type = encap_content_info.e_content_type;
+        let signed_attributes_content_type = signed_attributes
+            .iter()
+            .find(|attr| attr.oid.cmp(&PKCS9_CONTENT_TYPE_OID) == Ordering::Equal);
+        if let Some(signed_attributes_content_type) = signed_attributes_content_type {
+            // Check against `eContentType`
+            if signed_attributes_content_type.oid != e_content_type {
+                return Err(Error::Builder(
+                    "Mismatch between content types: encapsulated content info <-> signed attributes."
+                        .to_string(),
+                ));
+            }
+        } else {
+            signed_attributes.add(create_content_type_attribute(e_content_type)?)?;
+        }
+
+        // Now use `signer` to sign the DER encoded signed attributes
+        let mut signed_attributes_der = Vec::new();
+        signed_attributes.encode_to_vec(&mut signed_attributes_der)?;
+        let signature = signer.try_sign(&signed_attributes_der)?;
+        let signature_der = signature.to_vec();
+        self.signatures.push(signature_der);
+        let signature_der = self.signatures.last().expect("vector cannot be empty");
+        let signature_octetstringref = OctetStringRef::new(signature_der)?;
+        signer_info.signature = signature_octetstringref;
+        self.signer_infos.push(signer_info);
+
+        self.is_signed = true;
+
+        Ok(())
+    }
+
+    /// Return finalized `SignedData` struct.
+    /// This method returns a `ContentInfo` of type `signedData`. After this call, the builder cannot
+    /// be used any more. However, as the signature(s) are stored in the builder, the builder object
+    /// must be kept until the message is dropped.
+    pub fn build(&'s mut self) -> Result<ContentInfo<'s>> {
```

Comment:
Just wondering, would it make sense to reuse the [`Builder` trait](https://docs.rs/x509-cert/0.2.2-pre/x509_cert/builder/trait.Builder.html) from `x509-cert`, it's a dependency already.

Reply:
This would also hold for [`CertificateBuilder`](https://github.com/RustCrypto/formats/blob/bee8a6ea34f978ca474e4e42fceb158f9a39d6dd/x509-cert/src/builder.rs#L258), right?
As for the PKCS #7 builder: how would we handle multiple signers? My current solution is to add a `SignerInfo` including the signature with each call of `SignedDataBuilder::sign()`. Idea: I could possibly rename `sign()` to `add_signer()` and just add the `SignerInfo` without the signature. The signatures would then be calculated when building the message with `SignedDataBuilder::build()`.

Reply:
Yeah, I guess you would need an owned structure to be able to use it.

Reply:
@baloo I'm currently moving my implementations to the `cms` crate, which uses owned types. It looks like I will be able to reuse your `Builder` trait for the `SignerInfo` objects. 🙂